src/parquet/encodings/encoding-test.cc

```diff
@@ -0,0 +1,309 @@
+    ASSERT_EQ(expected[i], result[i]) << i;
+  }
+}
+
+template <>
+void VerifyResults<FLBA>(FLBA* result, FLBA* expected, int num_values) {
+  for (int i = 0; i < num_values; ++i) {
+    ASSERT_EQ(0, memcmp(expected[i].ptr, result[i].ptr, flba_length)) << i;
+  }
+}
+
+// ----------------------------------------------------------------------
+// Create some column descriptors
+
+template <typename T>
+std::shared_ptr<ColumnDescriptor> ExampleDescr() {
+  return nullptr;
+}
+
+template <>
+std::shared_ptr<ColumnDescriptor> ExampleDescr<FLBA>() {
+  auto node = schema::PrimitiveNode::MakeFLBA("name", Repetition::OPTIONAL,
+      flba_length, LogicalType::UTF8);
+  return std::make_shared<ColumnDescriptor>(node, 0, 0);
+}
+
+// ----------------------------------------------------------------------
+// Plain encoding tests
+
+template <typename Type>
+class TestPlainEncoding : public ::testing::Test {
+ public:
+  typedef typename Type::c_type T;
+  static constexpr int TYPE = Type::type_num;
+
+  void SetUp() {
+    descr_ = ExampleDescr<T>();
+  }
+
+  void InitData(int nvalues) {
+    num_values_ = nvalues;
+    input_bytes_.resize(num_values_ * sizeof(T));
+    output_bytes_.resize(num_values_ * sizeof(T));
+    draws_ = reinterpret_cast<T*>(input_bytes_.data());
+    decode_buf_ = reinterpret_cast<T*>(output_bytes_.data());
+  }
+
+  virtual void EncodeDecode() {
+    PlainEncoder<TYPE> encoder(descr_.get());
+    PlainDecoder<TYPE> decoder(descr_.get());
+
+    InMemoryOutputStream dst;
+    encoder.Encode(draws_, num_values_, &dst);
+
+    encode_buffer_ = dst.GetBuffer();
+
+    decoder.SetData(num_values_, encode_buffer_->data(),
+        encode_buffer_->size());
+    int values_decoded = decoder.Decode(decode_buf_, num_values_);
+    ASSERT_EQ(num_values_, values_decoded);
+  }
+
+  void Execute(int nvalues) {
+    InitData(nvalues);
+    GenerateData<T>(nvalues, draws_, &data_buffer_);
+    EncodeDecode();
+    VerifyResults<T>(decode_buf_, draws_, nvalues);
+  }
+
+ private:
+  int num_values_;
+  T* draws_;
+  T* decode_buf_;
+  vector<uint8_t> input_bytes_;
+  vector<uint8_t> output_bytes_;
+  vector<uint8_t> data_buffer_;
+  std::shared_ptr<Buffer> encode_buffer_;
+  std::shared_ptr<ColumnDescriptor> descr_;
+};
+
+TYPED_TEST_CASE(TestPlainEncoding, ParquetTypes);
+
+TYPED_TEST(TestPlainEncoding, BasicRoundTrip) {
+  this->Execute(10000);
+}
+
+// ----------------------------------------------------------------------
+// Dictionary encoding tests
+
+typedef ::testing::Types<Int32Type, Int64Type, Int96Type, FloatType, DoubleType,
+                         ByteArrayType, FLBAType> DictEncodedTypes;
+
+template <typename Type>
+class TestDictionaryEncoding : public ::testing::Test {
+ public:
+  typedef typename Type::c_type T;
+  static constexpr int TYPE = Type::type_num;
+
+  void SetUp() {
+    descr_ = ExampleDescr<T>();
+    if (descr_) {
+      type_length_ = descr_->type_length();
+    }
+  }
+
+  void InitData(int nvalues) {
```

Comment:
Can we avoid this duplication by creating a base `TestEncoding` class and deriving `PlainEncoding` and `DictionaryEncoding` from it ?